### PR TITLE
user_ldap: Filter groups after nested groups

### PR DESCRIPTION
**Pull request by Cybso**
Currently groupsMatchFilter is called before nested groups are resolved.
This basicly breaks this feature since it is not possible to inherit
membership in a group from another group.

Minimal example:

```
  Group filter: (&(objectClass=group),(cn=nextcloud))
  Nested groups: enabled

  cn=nextcloud,ou=Nextcloud,ou=groups,dn=company,dn=local
    objectClass: group

  cn=IT,ou=groups,dn=company,dn=local
    objectClass: group
    memberOf: cn=nextcloud,ou=Nextcloud,ou=groups,dn=company,dn=local

  cn=John Doe,ou=users,dn=company,dn=local
    objectClass: person
    memberOf: cn=IT,ou=groups,dn=company,dn=local
```

Since 'cn=IT,ou=groups,dn=company,dn=local' doesn't match the group
filter, John wouldn't be a member of group 'nextcloud'.

This patch fixes this by filtering the groups after all nested groups
have been collected. If nested groups is disabled the result will be the
same as without this patch.

Signed-off-by: Roland Tapken <roland@bitarbeiter.net>

### CODE REVIEW

**Comment by Cybso** [comment]
Sorry, this patch is incomplete. It only works for one level on indirection and fails if there is a second level:

```
  Group filter: (&(objectClass=group),(cn=nextcloud))
  Nested groups: enabled

  cn=nextcloud,ou=Nextcloud,ou=groups,dn=company,dn=local
    objectClass: group

  cn=IT,ou=groups,dn=company,dn=local
    objectClass: group
    memberOf: cn=nextcloud,ou=Nextcloud,ou=groups,dn=company,dn=local

  cn=Administrators,ou=groups,dn=company,dn=local
    objectClass: group
    memberOf: cn=IT,ou=groups,dn=company,dn=local

  cn=John Doe,ou=users,dn=company,dn=local
    objectClass: person
    memberOf:  cn=Administrators,ou=groups,dn=company,dn=local
```
I'll see if I can rewrite the patch and push it.

**Comment by Cybso** [comment]
The updated pull request is a more complex change, but it ensures that nested groups are resolved until their root.

**Comment by MorrisJobke** [comment]
@nextcloud/ldap Could you please have a look at this? It's already lying around for too long.

**Comment by blizzz** [comment]
> Currently groupsMatchFilter is called before nested groups are resolved.
This basicly breaks this feature since it is not possible to inherit
membership in a group from another group.

Yes, when the filter is as strict as that. With this changes, the filter does not apply to the subgroups so their members are being read despite that they are not whitelisted.

**Comment by Cybso** [comment]
@blizzz Thanks for your feedback. No, the filter is still applied to the subgroups, but only at the end after all sub groups have been resolved. This is the only way to allow the usage of distributions groups that are common in larger Active Directory configurations.

**Comment by blizzz** [comment]
@Cybso that's my point. It allows members of those groups to be retrieved while the groups themselves are not white listed. 

I am sure it makes sense from one point of view, but rather not from another and constitutes a behavioural change.

Further more, the better approach is perhaps to solely rely on  LDAP_MATCHING_RULE_IN_CHAIN so that the LDAP server resolves the the subgroups. I never tested it with OpenLDAP, however it would also be the better performing approach.

**Comment by Cybso** [comment]
@blizzz This is not for members of groups but for memberships of a user, and even if the intermediate groups are not listed the user **is** a member of the groups at the leaves.

LDAP_MATCHING_RULE_IN_CHAIN resolves the members of a group recursively, but in this case the opposite is required (get all groups where a person is member of, direct or indirect) and as far as I know there is now search syntax that provides this feature besides to recursively climb down the tree.

**Comment by blizzz** [comment]
> @blizzz This is not for members of groups but for memberships of a user, and even if the intermediate groups are not listed the user is a member of the groups at the leaves.

Ah, excuse me, I was mentally in another place of the code :speak_no_evil: 

> LDAP_MATCHING_RULE_IN_CHAIN resolves the members of a group recursively, but in this case the opposite is required (get all groups where a person is member of, direct or indirect) and as far as I know there is no search syntax that provides this feature besides to recursively climb down the tree.

True.

Can I ask also here for unit tests? TestCase offers ``invokePrivate()`` for calling restricted methods directly.

**Comment by MorrisJobke** [comment]
@blizzz Should we take it as it is or wait longer?

**Comment by Cybso** [comment]
I am very sorry, but at the moment I am extremely busy both professionally and privately, so I did not have the time to do that and probably will not find an opportunity to get into the test framework in the foreseeable future.

See also #8230

File: apps/user_ldap/lib/Group_LDAP.php [hunk 1]
```diff
@@ -250,29 +256,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+			$seen = array();
+			while ($group = array_pop($groups)) {
+				if ($group === $DN || array_key_exists($group, $seen)) {
+					// Prevent loops
+					continue;
+				}
+				$seen[$group] = 1;
+
+				// Resolve nested groups
+				if (isset($cachedNestedGroups[$group])) {
```
Comment (blizzz) [inline apps/user_ldap/lib/Group_LDAP.php]: this should be `$this->cachedNestedGroups[$group]`

File: apps/user_ldap/lib/Group_LDAP.php [hunk 2]
```diff
@@ -250,29 +256,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+					$nestedGroups = $cachedNestedGroups[$group];
```
Comment (blizzz) [inline apps/user_ldap/lib/Group_LDAP.php]: here as well

File: apps/user_ldap/lib/Group_LDAP.php [hunk 3]
```diff
@@ -250,29 +256,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+				} else {
+					$nestedGroups = $this->access->readAttribute($group, 'memberOf');
+					if (!is_array($nestedGroups)) {
+						$nestedGroups = [];
+					}
+					$cachedNestedGroups[$group] = $nestedGroups;
```
Comment (blizzz) [inline apps/user_ldap/lib/Group_LDAP.php]: and here

File: apps/user_ldap/lib/Group_LDAP.php [hunk 4]
```diff
@@ -250,29 +256,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+				}
+				foreach ($nestedGroups as $nestedGroup) {
```
Comment (blizzz) [inline apps/user_ldap/lib/Group_LDAP.php]: can be simplified with an array_merge if i am not mistaken

File: apps/user_ldap/lib/Group_LDAP.php [hunk 5]
```diff
@@ -250,29 +256,39 @@ private function _groupMembers($dnGroup, &$seen = null) {
...
+					array_push($groups, $nestedGroup);
+				}
 			}
+			// Get unique group DN's from those we have visited in the loop
+			$groups = array_keys($seen);
```
Comment (blizzz) [inline apps/user_ldap/lib/Group_LDAP.php]: this does not have any effect

**Review by blizzz — COMMENTED** [review]
I have some nitpicks but it is essentially done. I'd look into it when i am back. It's to the most degree also a bug fix, so should not be affected by the freeze for 16 imho.

**Comment by blizzz** [comment]
I took over, (extended) follow up PR is https://github.com/nextcloud/server/pull/14464

**Pull request closed without merging**